src/interface/swarm_device_context.hpp

```diff
@@ -0,0 +1,112 @@
+//========================================================================================
+// (C) (or copyright) 2021. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+#ifndef INTERFACE_SWARM_DEVICE_CONTEXT_HPP_
+#define INTERFACE_SWARM_DEVICE_CONTEXT_HPP_
+
+namespace parthenon {
+
+class SwarmDeviceContext {
+ public:
+  KOKKOS_FUNCTION
+  bool IsActive(int n) const { return mask_(n); }
+
+  KOKKOS_FUNCTION
+  bool IsOnCurrentMeshBlock(int n) const { return blockIndex_(n) == this_block_; }
+
+  KOKKOS_FUNCTION
+  void MarkParticleForRemoval(int n) const { marked_for_removal_(n) = true; }
+
+  KOKKOS_FUNCTION
+  bool IsMarkedForRemoval(const int n) const { return marked_for_removal_(n); }
+
+  KOKKOS_INLINE_FUNCTION
+  int GetNeighborBlockIndex(const int &n, const double &x, const double &y,
+                            const double &z, bool &is_on_current_mesh_block) const {
+    int i = static_cast<int>(std::floor((x - x_min_) / ((x_max_ - x_min_) / 2.))) + 1;
+    int j = static_cast<int>(std::floor((y - y_min_) / ((y_max_ - y_min_) / 2.))) + 1;
+    int k = static_cast<int>(std::floor((z - z_min_) / ((z_max_ - z_min_) / 2.))) + 1;
+
+    // Something went wrong
+    if (i < 0 || i > 3 || ((j < 0 || j > 3) && ndim_ > 1) ||
+        ((k < 0 || k > 3) && ndim_ > 2)) {
+      PARTHENON_FAIL("Particle neighbor indices out of bounds");
+    }
+
+    // Ignore k,j indices as necessary based on problem dimension
+    if (ndim_ == 1) {
+      blockIndex_(n) = neighborIndices_(0, 0, i);
+    } else if (ndim_ == 2) {
+      blockIndex_(n) = neighborIndices_(0, j, i);
+    } else {
+      blockIndex_(n) = neighborIndices_(k, j, i);
+    }
+
+    is_on_current_mesh_block = (blockIndex_(n) == this_block_);
+
+    return blockIndex_(n);
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  int GetMyRank() const { return my_rank_; }
+
+  KOKKOS_INLINE_FUNCTION
+  void Xtoijk(const Real &x, const Real &y, const Real &z, int &i, int &j, int &k) const {
```

Comment:
Following up on my earlier comment: This function could be templated (or otherwise safeguarded) for use only with Cartesian coordinates.

Reply:
Yes when we go beyond uniform cartesian grids this will need to be redesigned... but depending on what profiling looks like this approach to the swarm device context may need to be redesigned first anyway. 

A safeguard would be good to add now, I just added 
```c++
  PARTHENON_REQUIRE_THROWS(typeid(Coordinates_t) == typeid(UniformCartesian),                        
    "SwarmDeviceContext only supports a uniform Cartesian mesh!");
```
to the `Swarm` constructor, does that look good to you? (Can't use `typeid` in device code i.e. inside `SwarmDeviceContext` calls)

Reply:
Yes, I think a single check early on is fine.
Potentially you could add comments to the places in the code that'd need to be changed (right now, after you wrote the code they may be more obvious than a couple to hours/days/weeks/... down the road when make/need that change), but it's just a suggestion.

Reply:
months/years/... :) Yes, good idea, I've added some TODOs in the relevant parts of `SwarmDeviceContext` for later on.